plutus-metatheory/src/Signature.lagda.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Signatures
+layout: page
+---
+
+Module for abstract signatures of builtins:
+
+This module defines signatures for characterizing the types of built-in functions.
+
+The signatures defined in this module are abstract in the sense 
+that they don't refer to any particular typing context or typing rule,
+except for the minimal rule of guaranteeing well-formedness.
+
+## Imports 
+
+```
+module Signature where
+
+open import Data.Nat using (ℕ;zero;suc)
+open import Data.Fin renaming (zero to Z; suc to S)
+open import Data.List.NonEmpty using (List⁺;foldr)
+
+open import Utils using (*)
+open import Type using (Ctx⋆;∅;_,⋆_;_⊢⋆_;_∋⋆_;Z;S;Φ)
+```
+
+## Built-in compatible types 
+
+The arguments of a built-in function can't be just any type, but are restricted 
+to certain types, which in this file are called *built-in-compatible* types.
+
+The built-in compatible types are either type constants, type variables, 
+or type operators applied to built-in-compatible type.
+
+The type of built-in-compatible types (_⊢♯) is indexed by the number of 
+distinct type variables.
+```
+data _⊢♯ : ℕ → Set
+
+open import Builtin.Constant.Type ℕ (_⊢♯) using (TyCon)
+
+data _⊢♯ where
+  -- a type variable
+  ` : ∀ {n} → 
+      Fin n
+      --------
+    → n ⊢♯
+
+  -- a type constant or type operator applied to a built-in-compatible type
+  con : ∀ {n} → 
+      TyCon n
+      ------
+      → n ⊢♯
+```
+
+The list of arguments is a non-empty list of built-in compatible types.
+It is indexed by the number of distinct type variables that may appear.
+
+```
+
+Args : ℕ → Set
+Args n = List⁺ (n ⊢♯) 
+
+```
+
+## Signatures
+
+A signature is given by
+  1. The number (fv♯) of type variables that may appear
+  2. A list of arguments
+  3. A result type
+
+```
+record Sig : Set where 
+   constructor sig
+   field
+     -- number of type variables
+    fv♯ : ℕ   
+     -- list of arguments
+    args : Args fv♯
+     -- type of result
+    result : fv♯ ⊢♯
+
+```
+
+## Obtaining concrete types from signatures
+
+Signatures represent abstract types which need to be made concrete in 
+order to use them. The following modules may be instantiated to obtain 
+a function `sig2type` which converts from an abstract into a concrete type.
+
+The instantation is defined in two stages. In the first stage, the following parameters are instantiated:
```

Comment:
Why does it need to be two-staged?

Reply:
I'm also interested in this

Reply:
Because I was too stupid to figure out how to do it in one stage. It's  done in one stage now.